winch/codegen/src/isa/x64/masm.rs

```diff
@@ -2543,6 +2543,183 @@ impl Masm for MacroAssembler {
         Ok(())
     }
 
+    fn v128_trunc_sat(
+        &mut self,
+        context: &mut CodeGenContext<Emission>,
+        kind: V128TruncSatKind,
+    ) -> Result<()> {
+        self.ensure_has_avx()?;
+
+        let reg = writable!(context.pop_to_reg(self, None)?.reg);
+        let scratch = writable!(regs::scratch_xmm());
+
+        match kind {
+            V128TruncSatKind::F32x4S => {
+                // Create a mask to handle NaN values (1 for not NaN, 0 for
+                // NaN).
+                self.asm.xmm_vcmpp_rrr(
+                    scratch,
+                    reg.to_reg(),
+                    reg.to_reg(),
+                    kind.src_lane_size(),
+                    VcmpKind::Eq,
+                );
+                // Zero out any NaN values.
+                self.asm
+                    .xmm_vandp_rrr(reg.to_reg(), scratch.to_reg(), reg, kind.src_lane_size());
+                // Create a mask for the sign bits.
+                self.asm
+                    .xmm_vex_rr(AvxOpcode::Vpxor, scratch.to_reg(), reg.to_reg(), scratch);
+                // Convert floats to integers.
+                self.asm.xmm_vcvt_rr(reg.to_reg(), reg, VcvtKind::F32ToI32);
+                // Apply sign mask to the converted integers.
+                self.asm
+                    .xmm_vex_rr(AvxOpcode::Vpand, reg.to_reg(), scratch.to_reg(), scratch);
+                // Create a saturation mask of all 1s for negative numbers,
+                // all 0s for positive numbers. The arithmetic shift will cop
+                // the sign bit.
+                self.asm
+                    .xmm_vpsra_rri(scratch.to_reg(), scratch, 0x1F, kind.dst_lane_size());
+                // Combine converted integers with saturation mask.
+                self.asm
+                    .xmm_vex_rr(AvxOpcode::Vpxor, reg.to_reg(), scratch.to_reg(), reg);
+            }
+            V128TruncSatKind::F32x4U => {
+                let reg2 = writable!(context.any_fpr(self)?);
```

Comment:
I'm confused about the lifetime of this register, I was expecting a `context.free_reg` given that it seems to be temporary?

Reply:
Good catch!